core/src/main/java/org/elasticsearch/common/settings/SettingsService.java

```diff
@@ -0,0 +1,201 @@
+ * under the License.
+ */
+
+package org.elasticsearch.common.settings;
+
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+/**
+ * A basic setting service that can be used for per-index and per-cluster settings.
+ * This service offers transactional application of updates settings.
+ */
+public abstract class SettingsService extends AbstractComponent {
+    private Settings lastSettingsApplied;
+    private final List<SettingUpdater> settingUpdaters = new ArrayList<>();
+
+    protected SettingsService(Settings settings) {
+        super(settings);
+    }
+
+    /**
+     * Applies the given settings to all listeners and rolls back the result after application. This
+     * method will not change any settings but will fail if any of the settings can't be applied.
+     */
+    public synchronized Settings dryRun(Settings settings) {
+        final Settings build = Settings.builder().put(this.settings).put(settings).build();
+        try {
+            List<RuntimeException> exceptions = new ArrayList<>();
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (RuntimeException ex) {
+                    exceptions.add(ex);
+                    logger.debug("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                }
+            }
+            // here we are exhaustive and record all settings that failed.
+            ExceptionsHelper.rethrowAndSuppress(exceptions);
+        } finally {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.rollback();
+                } catch (Exception e) {
+                    logger.warn("failed to rollback settings for [{}]", e, settingUpdater);
+                }
+            }
+        }
+        return build;
+    }
+
+    /**
+     * Applies the given settings to all the settings consumers or to none of them. The settings
+     * will be merged with the node settings before they are applied while given settings override existing node
+     * settings.
+     * @param settings the settings to apply
+     * @return the unmerged applied settings
+    */
+    public synchronized Settings applySettings(Settings settings) {
+        if (lastSettingsApplied != null && settings.equals(lastSettingsApplied)) {
+            // nothing changed in the settings, ignore
+            return settings;
+        }
+        final Settings build = Settings.builder().put(this.settings).put(settings).build();
```

Comment:
Maybe rename the parameter to `newSettings`? In general I don't like the name shadowing unless its in a setter or something obvious.